src/components/viewmodels/rooms/UserIdentityWarningViewModel.tsx

```diff
@@ -0,0 +1,164 @@
+                } else {
+                    candidatePrompt = undefined;
+                }
+
+                // is the current prompt still valid?
+                setCurrentPrompt((existingPrompt): ViolationPrompt | undefined => {
+                    if (existingPrompt && violations.includes(existingPrompt)) {
+                        return existingPrompt;
+                    } else if (candidatePrompt) {
+                        return candidatePrompt;
+                    } else {
+                        return undefined;
+                    }
+                });
+            }),
+        [crypto, room],
+    );
+
+    // We need to listen for changes to the members list
+    const onRoomStateEvent = useCallback(
+        async (event: MatrixEvent): Promise<void> => {
+            if (!crypto || event.getRoomId() !== room.roomId) {
+                return;
+            }
+            let shouldRefresh = false;
+
+            const eventType = event.getType();
+
+            if (eventType === EventType.RoomEncryption && event.getStateKey() === "") {
+                // Room is now encrypted, so we can initialise the component.
+                shouldRefresh = true;
+            } else if (eventType == EventType.RoomMember) {
+                // We're processing an m.room.member event
+                // Something has changed in membership, someone joined or someone left or
+                // someone changed their display name. Anyhow let's refresh.
+                const userId = event.getStateKey();
+                shouldRefresh = !!userId;
+            }
+
+            if (shouldRefresh) {
+                loadViolations().catch((e) => {
+                    logger.error("Error refreshing UserIdentityWarningViewModel:", e);
+                });
+            }
+        },
+        [crypto, room, loadViolations],
+    );
+    useTypedEventEmitter(cli, RoomStateEvent.Events, onRoomStateEvent);
+
+    // We need to listen for changes to the verification status of the members to refresh violations
+    const onUserVerificationStatusChanged = useCallback(
+        (userId: string): void => {
+            if (members.find((m) => m.userId == userId)) {
+                // This member is tracked, we need to refresh.
+                // refresh all for now?
+                // As a later optimisation we could store the current violations and only update the relevant one.
+                loadViolations().catch((e) => {
+                    logger.error("Error refreshing UserIdentityWarning:", e);
+                });
+            }
+        },
+        [loadViolations, members],
+    );
+    useTypedEventEmitter(cli, CryptoEvent.UserTrustStatusChanged, onUserVerificationStatusChanged);
+
+    useEffect(() => {
+        loadViolations().catch((e) => {
+            logger.error("Error initialising UserIdentityWarning:", e);
+        });
+    }, [loadViolations]);
+
+    let onButtonClick: (ev: ButtonEvent) => void = () => {};
```

Comment:
It's a bit weird to assign a different callback to the button depending on the state. Generally you would just have one callback which behaves accordingly. Also, you probably want to make it a `useCallback` and ideally be a bit more specific than onButtonClick (even though in this case there's only one button).

Reply:
Thx, I tried that https://github.com/element-hq/element-web/pull/29067/commits/d1509b96304a87bd7ffd718be4d5a1ec672f8146